### PR TITLE
fix: order of evaluation for some builtins

### PR DESCRIPTION
**Pull request by charles-cooper**
ecadd, ecmul, addmod, mulmod

in the case that the arguments have side effects, they could be evaluated out of order

chainsec june 2023 review 5.1

patch for https://github.com/vyperlang/vyper/security/advisories/GHSA-4hg4-9mf5-wxxq

### Commit message

Commit message for the final, squashed PR. (Optional, but reviewers will appreciate it! Please see [our commit message style guide](../../blob/master/docs/style-guide.rst#best-practices-1) for what we would ideally like to see in a commit message.)

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]

File: vyper/builtins/functions.py [hunk 1]
```diff
@@ -824,27 +814,22 @@ class ECMul(BuiltinFunction):
...
+
+        ret = ["seq"]
+
+        dst0 = IRnode.from_list(buf, typ=SArrayT(UINT256_T, 2), location=MEMORY)
+        ret.append(make_setter(dst0, args[0]))
+
+        dst1 = IRnode.from_list(buf + 64, typ=UINT256_T, location=MEMORY)
+        ret.append(make_setter(dst1, args[1]))
+
+        ret.append(["assert", ["staticcall", ["gas"], 7, buf, 128, buf, 64]])
```
Comment (trocher) [inline vyper/builtins/functions.py:829]: Shouldn't the precompile only take 96 bytes of input ?
```suggestion
        ret.append(["assert", ["staticcall", ["gas"], 7, buf, 96, buf, 64]])
```

Comment (charles-cooper) [reply vyper/builtins/functions.py:829]: Ah yea, seems I changed the buffer size but not the call input

**Comment by pcaversaccio** [comment]
Point for a quick discussion: do you guys see any need to document this (or link to [GHSA-4hg4-9mf5-wxxq](https://github.com/vyperlang/vyper/security/advisories/GHSA-4hg4-9mf5-wxxq)) in the docs?

**Comment by pcaversaccio** [comment]
As per offline discussion with @charles-cooper, it would make sense to have a general bug section in the docs; similar to https://docs.soliditylang.org/en/v0.8.21/bugs.html. I will open an issue accordingly.

**Pull request merged**